### PR TITLE
Optional time scales

### LINKED ISSUE #112: Make day and date variables optional
Chart should be able to be drawn with *either* day or date variables (as specified by user in settings). Having both days and dates is fine, but not required. 

# Test notes
* Confirm that the tools works when there are missing values for either day or date variables
  1. Open "clinical-timeline" using the latest commit from the "optional-timescales" branch and the "testData/ADTIMELINES_partialMissing.csv" dataset. 
  2. Note that there are 10 participants with no "Day" info, and 5 participants with no "Date" info in this data set. 
  3. Confirm that there are 145 participants when x-axis type = Date and 140 when X-axis scale = day.

### PR DESCRIPTION
**Pull request by samussiah**
Optional-izing the time scale required a significant refactor.

Closes #112

### CODE REVIEW

File: src/callbacks/onLayout.js [hunk 1]
```diff
@@ -1,65 +1,17 @@
...
-                .text('Controls');
-        else if (d.option === 'y.grouping') {
-            const filterRule = context.controls.wrap
-                .append('div')
-                .classed('filters horizontal-rule', true)
-                .text('Filters');
-            context.controls.wrap.node().insertBefore(filterRule.node(), this.nextSibling);
-        }
-    });
+    //Lay out details container.
```
Comment (jwildfire) [inline src/callbacks/onLayout.js]: Really nice job modularizing all of this.

File: src/callbacks/onLayout/augmentOtherControls/timeScaleChange.js [hunk 2]
```diff
@@ -9,28 +10,15 @@ export default function timeScaleChange(dropdown, d) {
...
-    this.listing.config.headers = this.listing.config.headers.map(header => {
-        if (header === 'Start Day') header = 'Start Date';
-        else if (header === 'Start Date') header = 'Start Day';
-        if (header === 'Stop Day') header = 'Stop Date';
-        else if (header === 'Stop Date') header = 'Stop Day';
-
-        return header;
-    });
+    //Remove records without time data.
+    cleanData.call(this);
```
Comment (jwildfire) [inline src/callbacks/onLayout/augmentOtherControls/timeScaleChange.js]: Lets say there are 10 missing days and 20 missing dates in a 100 record data set (no overlap). Will there be 90 records in one view, and 80 records in the other - or will there be 70 records in both views? Or some other configuration?

Comment (jwildfire) [reply src/callbacks/onLayout/augmentOtherControls/timeScaleChange.js]: Let's add a test case that explicitly spells out (and then tests) the expected behavior here.

Comment (samussiah) [reply src/callbacks/onLayout/augmentOtherControls/timeScaleChange.js]: cleanData() is called after updating the time scale so it removes missing values in that variable from the initial raw data.

File: src/callbacks/onLayout.js (hunk 1 above)
Comment (samussiah) [reply src/callbacks/onLayout.js]: Thanks!  Makes it a lot easier to follow the sequence of events.

File: src/callbacks/onLayout/augmentOtherControls/timeScaleChange.js (hunk 2 above)
Comment (jwildfire) [reply src/callbacks/onLayout/augmentOtherControls/timeScaleChange.js]: Gotcha - so in the case above we're expecting 70 records in both views?

**Comment by jwildfire** [comment]
Adding quick fix for #127.

File: src/callbacks/onLayout/augmentOtherControls/timeScaleChange.js (hunk 2 above)
Comment (samussiah) [reply src/callbacks/onLayout/augmentOtherControls/timeScaleChange.js]: No, 90 day records and 80 date records.

**Comment by dschwentker** [comment]
https://github.com/RhoInc/clinical-timelines/issues/112 was tested, passed, and closed.  https://github.com/RhoInc/clinical-timelines/issues/127 was already closed and required no testing, so this PR has been fully tested.

**Pull request merged**